### PR TITLE
Add cache tests

### LINKED ISSUE #185: Write test for behavior of CachedBInding's trySetMember
In the `Cached` case of `trySetMember`, the assignment of `None` (as opposed to `Some value`) to the field `Cache` is correct.  As pointed out in https://github.com/elmish/Elmish.WPF/pull/183#discussion_r344018327, assigning `None` allows format-as-you-type behavior.  We should create a test that depends on that behavior.
 https://github.com/elmish/Elmish.WPF/blob/7feeb9bc23e38334ea41bb36c2fbf334b7b04ba0/src/Elmish.WPF/ViewModel.fs#L698-L702

### PR DESCRIPTION
**Pull request by TysonMN**
Resolves #185

### CODE REVIEW

File: src/Elmish.WPF.Tests/DynamicViewModelTests.fs [hunk 1]
```diff
@@ -1410,6 +1410,75 @@ module SubModelSelectedItem =
...
+    let binding =
+      name
+      |> Binding.TwoWay.id
+      |> Binding.mapModel mapping.Fn
+      |> Binding.addCaching
+    let vm = TestVm(initialModel, binding)
+
+    vm.Get name |> ignore // populate cache
+    vm.Set name newModel
+    mapping.Reset()       // Set calls mapping function, so reset count
```
Comment (TysonMN) [inline src/Elmish.WPF.Tests/DynamicViewModelTests.fs:1474]: @marner2, I don't think this caching effect can prevent the model mapping function from being called on `Set` when the cache is populated.  `Set` needs the model but the cache contains a different type.

Do you agree?

File: src/Elmish.WPF.Tests/DynamicViewModelTests.fs [hunk 2]
```diff
@@ -1410,6 +1410,75 @@ module SubModelSelectedItem =
 
 
 
+module CacheEffect =
+
+  [<Fact>]
+  let ``model mapping called excactly once when Get called twice`` () =
```
Comment (the-not-mad-psychologist) [inline src/Elmish.WPF.Tests/DynamicViewModelTests.fs:1416]: *exactly

File: src/Elmish.WPF/BindingVmHelpers.fs [hunk 3]
```diff
@@ -699,10 +699,9 @@ type Set(value: obj) =
     match binding with
     | BaseVmBinding b -> this.Base(model, b)
     | Cached b ->
-        let successful = this.Recursive(model, b.Binding)
-        if successful then
-          b.Cache.Value <- None  // TODO #185: write test
-        successful
+        // Set only dispatches a message and while UpdateModel changes the model,
+        // so don't clear the cache here
```
Comment (the-not-mad-psychologist) [inline src/Elmish.WPF/BindingVmHelpers.fs:703]: Grammatically unclear. "While" would work, but "and while" doesn't when used like that. Maybe also add a period at the end.

Comment (the-not-mad-psychologist) [reply src/Elmish.WPF/BindingVmHelpers.fs:703]: Or do you mean `while UpdateModel`? I'm assuming you're not using a `while`-loop, though.

Comment (TysonMN) [reply src/Elmish.WPF/BindingVmHelpers.fs:704]: Thanks for the review.

I reworded. Is that clear now?

File: src/Elmish.WPF.Tests/DynamicViewModelTests.fs (hunk 1 above)
Comment (marner2) [reply src/Elmish.WPF.Tests/DynamicViewModelTests.fs:1474]: I'm fighting very hard against this, but I think you're right. "Set" is really "DispatchMessage." The model isn't affected until you actually run "Update."

I agree.

Comment (TysonMN) [reply src/Elmish.WPF.Tests/DynamicViewModelTests.fs:1474]: Can you elaborate on what you mean by

> I'm fighting very hard against this

?

Comment (marner2) [reply src/Elmish.WPF.Tests/DynamicViewModelTests.fs:1474]: It just seems wrong to not bust the cache when you set it. But I think I'm pretty convinced that the "actual" set happens on `Update`.

File: src/Elmish.WPF/BindingVmHelpers.fs (hunk 3 above)
Comment (the-not-mad-psychologist) [reply src/Elmish.WPF/BindingVmHelpers.fs:704]: Much better, thanks!

**Review by the-not-mad-psychologist — APPROVED** [review]

**Comment by marner2** [comment]
Maybe as part of another PR. I think it’s clear enough here.

File: src/Elmish.WPF.Tests/DynamicViewModelTests.fs (hunk 1 above)
Comment (marner2) [reply src/Elmish.WPF.Tests/DynamicViewModelTests.fs:1474]: Maybe as part of another PR. I think it’s clear enough here.

**Pull request merged**